### ISSUE TITLE
Support for questions asked by git or ssh

### ISSUE THREAD

**Issue by alexreg** [theseal/ssh-askpass#14]
When the dialog appears, there is no textbox to enter the username/password. It looks like the following:

![screenshot](https://cldup.com/uDBuYn0vgbh/pYOEEq.png-thumb.png)

**Comment by simmel** [comment]
Looks like you have cloned an `https://github.com/...`-repo. Using
`https` with ssh-ask pass isn't something that's currently supported. We
might support it if more people ask for it because you are the first = )

In the mean time, use the git-over-ssh. If you haven't made any changes
to your repo yet, just delete it and reclone it see
https://help.github.com/articles/which-remote-url-should-i-use/#cloning-with-ssh-urls

If you have commits please see this article on how to change your origin
remote URL http://superuser.com/q/199507

**Comment by alexreg** [comment]
Ah, I didn't realise that. I must have not explicitly specified the Git-SSH remote when pushing. Thanks for clarifying.

And yes, it would be a nice feature. Please do consider it. :)

**Comment by ZanYzer** [comment]
I had a similar issue where there was no textbox to enter the passphrase for an SSH key.

![screenshot-askpass-dialog](https://cloud.githubusercontent.com/assets/14753152/24405351/e0698bb4-13c4-11e7-9f18-20ad02976b30.png)

I'm using git-over-ssh (git@bitbucket.org). For more details see my [my StackOverflow question](http://stackoverflow.com/questions/43067233/checkout-a-git-submodule-in-xcode-8-build-phase-script).

**Comment by simmel** [comment]
As a work around you can run `ssh-add -c` in a Terminal before using Xcode @ZanYzer, does that solve it?

**Comment by ZanYzer** [comment]
After adding my private key with `ssh-add -c` i got a new error:

`sign_and_send_pubkey: signing failed: agent refused operation`

I did some reading and [found that this has something to do with my keyring](http://askubuntu.com/questions/762541/ubuntu-16-04-ssh-sign-and-send-pubkey-signing-failed-agent-refused-operation) and that by adding SSH_AUTH_SOCK=0 to the beginning of the script i can test this. I added that line and got the exact same dialog box i posted earlier.

Also, according to a comment on the linked question above ssh-agent automatically detects files like `~/.ssh/id_rsa` and `~/.ssh/id_rsa.pub` and that seems to be true so there's no need to manually add those files.

**Comment by ZanYzer** [comment]
I used `ssh-add -d` to remove the identity i just added to the agent, removed SSH_AUTH_SOCK=0 from the script, used `ssh-add -l` and `ssh-add -L` to list the identities (both returned "The agent has no identities.") and tried runnig the script again. Same result, same dialog box, so indeed the ssh-agent does automatically manage `~/.ssh/id_rsa` and `~/.ssh/id_rsa.pub`, so adding those manually just causes unexpected problems.

**Comment by simmel** [comment]
I have no idea what's wrong. `sign_and_send_pubkey: signing failed: agent refused operation` is what `ssh` says when you press cancel in `ssh-askpass` or it can't execute other wise AFAIK.

Maybe there are two ssh-agent's running on your system? Have you tried to reboot? What does `ps auxww | grep ssh-agent` say when you're getting that dialog via Xcode?

**Comment by ZanYzer** [comment]
Right after reboot:
```
USER$ ps auxww | grep ssh-agent
USER               500   0.0  0.0  2432804   2008 s000  S+    4:46PM   0:00.00 grep ssh-agent
```
I would say this is expected since i didn't configure agents to launch with the system.

After running the script and getting the broken dialog box:
```
USER$ ps auxww | grep ssh-agent
USER               714   0.0  0.0  2432804   2008 s000  S+    4:53PM   0:00.00 grep ssh-agent
USER               676   0.0  0.1  2435588   6180   ??  S     4:50PM   0:00.03 /usr/bin/ssh-agent -l
```
As expected the agent starts running automatically and there's only one.

By the way, when i press cancel on the broken dialog box this is the error i receive in Xcode:
![screenshot-user-cancelled](https://cloud.githubusercontent.com/assets/14753152/24461123/846e183e-14a0-11e7-9144-ef85de7eabc7.png)
Xcode even opens the ssh-askpass script unlike when it fails with the `sign_and_send_pubkey:` error.

---

Just to be super clear:

When i run the same script from the command line (and not from Xcode) i'm prompted for the passphrase right there and everything works as expected:
![screenshot-prompt-command-line](https://cloud.githubusercontent.com/assets/14753152/24461714/3ed64a4c-14a2-11e7-91bd-8f829808d14e.png)

When i use Xcode's interface to do a checkout i'm prompted by Xcode for an authentication method and i have to switch it to "SSH Keys" and type in my passphrase, but because i've already allowed it to save this preference and the passphrase to the keychain it actually doesn't even prompt me anymore. Xcode's dialog box looks like this (ignore the Address, just found this image online):
![illustration-prompt-xcode](https://i.stack.imgur.com/813wg.png)

The problem only arises when the script is run by Xcode in a Build Phase. Probably that's why it's looking for `ssh-askpass` in such a strange location (`/usr/X11R6/bin` which is just a symlink to `/opt/X11` created by ???)

Sorry for the long post, [here's a talking potato](http://orig02.deviantart.net/d2e7/f/2012/126/7/7/glados_quote_potato_by_nathanr2013-d4yqney.jpg).

**Comment by theseal** [comment]
If you add the key to the ssh-agent **before** you start Xcode, could you get the environment from the ssh-agent so that we can verify that it is "our" agent that is running and not the built-in alternative:
```
$ ssh-add -c
Enter passphrase for /Users/seal/.ssh/id_rsa (will confirm each use):
Identity added: /Users/seal/.ssh/id_rsa (/Users/seal/.ssh/id_rsa)
The user must confirm each use of the key
$ ps -wwE -p $(pgrep ssh-agent)
15885 ??         0:00.55 /usr/bin/ssh-agent -l TMPDIR=/var/folders/k_/kaka SHELL=/bin/bash HOME=/Users/seal SSH_ASKPASS=/usr/local/opt/ssh-askpass/bin/ssh-askpass Apple_PubSub_Socket_Render=/private/tmp/com.apple.launchd.kaka/Render SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.kaka/Listeners LOGNAME=seal PATH=/usr/bin:/bin:/usr/sbin:/sbin DISPLAY=0 XPC_SERVICE_NAME=homebrew.mxcl.ssh-askpass USER=seal XPC_FLAGS=0x9
```
If `SSH_ASKPASS` isn't set to "/usr/local/opt/ssh-askpass/bin/ssh-askpass" you are not using our ssh-askpass and you should verify with `brew services list` that ssh-askpass is started.

**Comment by ZanYzer** [comment]
Just as you asked:
```
$ cd ~/.ssh
$ ssh-add -c id_rsa
Enter passphrase for id_rsa (will confirm each use): 
Identity added: id_rsa (id_rsa)
The user must confirm each use of the key
$ ps -wwE -p $(pgrep ssh-agent)
  PID TTY           TIME CMD
 1072 ??         0:00.04 /usr/bin/ssh-agent -l TMPDIR=/var/folders/sf/_48y5rz13t7d1h02w19pwjbc0000gn/T/ SHELL=/bin/bash HOME=/Users/MyUserName Apple_PubSub_Socket_Render=/private/tmp/com.apple.launchd.hJVBw8FkJB/Render SSH_AUTH_SOCK=/private/tmp/com.apple.launchd.aJonIK5ybq/Listeners LOGNAME=MyUserName PATH=/usr/bin:/bin:/usr/sbin:/sbin DISPLAY=/private/tmp/com.apple.launchd.ZuLUKPs7FZ/org.macosforge.xquartz:0 XPC_SERVICE_NAME=com.openssh.ssh-agent USER=MyUserName XPC_FLAGS=0x9
```

Looks like it's not set and `XPC_SERVICE_NAME ` doesn't match yours either.
`brew services list` gave me this:
```
$ brew services list
Name        Status  User Plist
ssh-askpass stopped
```

I installed ssh-askpass with Homebrew and received no warnings or errors during installation so i thought everything is set up. Your Installation/Homebrew doesn't state any additional steps either. Will `cp ssh-askpass.plist ~/Library/LaunchAgents/` and a reboot solve this or do i need to export SSH_ASKPASS into PATH manually? Are there any other steps i'm missing to make this work?

**Comment by ZanYzer** [comment]
I can confirm that just by adding the identity with `ssh-add -c` **before** Xcode is started resolved my issue. Now i'm getting the prompt from your script when i run the Build Phase script **and** when i run the script from the command line. `brew services list` still says that ssh-askpass is stopped, is that to be expected?

**Comment by simmel** [comment]
If you did `cp ssh-askpass.plist ~/Library/LaunchAgents/` then yes it's
to be expected.

**Comment by bcg62** [comment]
+1 for the original use case. If you use ssh-keys with a password or pin, in my case smartcards/pkcs11 you need to be able to enter text. This would also allow the script to be used for SUDO_ASKPASS

**Comment by alexreg** [comment]
Any chance of this feature being added yet? :-)

**Comment by simmel** [comment]
Can everyone try the new tag? I've updated the tap.

**Comment by bcg62** [comment]
@simmel tested with `ssh-add -e` and `SUDO_ASKPASS=/usr/local/bin/sudo-askpass sudo -A` both prompted and accepted my pin/password.

Thanks!

**Comment by bcg62** [comment]
It appears this is because the sudo prompt is `Password:` while you are checking for `pass`

**Comment by simmel** [comment]
Ah, cool! I didn't know sudo supported askpass!

"contains" in AppleScript is case insensitive so it works for both
"passphrase" and "Password" but the issue was that sudo asks for
"Password:" (note no trailing whitespace) and ssh asks for "Password: "
(note trailing whitespace). I added so we also check for prompts without
trailing space. I suspected that this would happen but didn't want to do
anything before you all checked.

I'll release a new tag with the fix in a few minutes.

Thank you for your testing Brendan!
